maze_transformer/generation/lattice_maze.py

```diff
@@ -306,17 +364,426 @@ def from_tokens(cls, tokens: list[str]) -> "LatticeMaze":
+            solution.append(tuple(neighbors_filtered[0]))
+
+        # assert the solution is complete
+        assert solution[0] == tuple(
+            start_pos
+        ), f"solution {solution} does not start at start_pos {start_pos}"
+        assert solution[-1] == tuple(
+            end_pos
+        ), f"solution {solution} does not end at end_pos {end_pos}"
+
+        return cls(
+            connection_list=np.array(connection_list),
+            solution=np.array(solution),
+        )
+
+    # ============================================================
+    # to and from ASCII
+    # ============================================================
+    def _as_ascii_grid(self) -> Shaped[np.ndarray, "x y"]:
+        # Get the pixel grid using to_pixels().
+        pixel_grid: Bool[np.ndarray, "x y"] = self._as_pixels_bw()
+
+        # Replace pixel values with ASCII characters.
+        ascii_grid: Shaped[np.ndarray, "x y"] = np.full(
+            pixel_grid.shape, AsciiChars.WALL, dtype=str
+        )
+        ascii_grid[pixel_grid == True] = AsciiChars.OPEN
+
+        return ascii_grid
+
+    def as_ascii(
+        self,
+        show_endpoints: bool = True,
+        show_solution: bool = True,
+    ) -> str:
+        """return an ASCII grid of the maze"""
+        ascii_grid: Shaped[np.ndarray, "x y"] = self._as_ascii_grid()
+        pixel_grid: PixelGrid = self.as_pixels(
+            show_endpoints=show_endpoints, show_solution=show_solution
+        )
+
+        chars_replace: tuple = tuple()
+        if show_endpoints:
+            chars_replace += (AsciiChars.START, AsciiChars.END)
+        if show_solution:
+            chars_replace += (AsciiChars.PATH,)
+
+        for ascii_char, pixel_color in ASCII_PIXEL_PAIRINGS.items():
+            if ascii_char in chars_replace:
+                ascii_grid[(pixel_grid == pixel_color).all(axis=-1)] = ascii_char
+
+        return "\n".join("".join(row) for row in ascii_grid)
+
+    @classmethod
+    def from_ascii(cls, ascii_str: str) -> "LatticeMaze":
+        lines: list[str] = ascii_str.strip().split("\n")
+        ascii_grid: Shaped[np.ndarray, "x y"] = np.array(
+            [list(line) for line in lines], dtype=str
+        )
+        pixel_grid: PixelGrid = np.zeros((*ascii_grid.shape, 3), dtype=np.uint8)
+
+        for ascii_char, pixel_color in ASCII_PIXEL_PAIRINGS.items():
+            pixel_grid[ascii_grid == ascii_char] = pixel_color
+
+        return cls.from_pixels(pixel_grid)
+
+
+@serializable_dataclass(frozen=True, kw_only=True)
+class TargetedLatticeMaze(LatticeMaze):
+    """A LatticeMaze with a start and end position"""
+
+    # this jank is so that SolvedMaze can inherit from this class without needing arguments for start_pos and end_pos
+    start_pos: Coord
+    end_pos: Coord
+
+    def __post_init__(self) -> None:
+        # make things numpy arrays (very jank to override frozen dataclass)
```

Comment:
If you need to do this jank, maybe it's a sign that this shouldn't be a frozen dataclass?